File: core/src/identity/keystore_file.rs
//! Filesystem-backed `KeyStore` implementation.
//!
//! Reads and writes `~/.mnemonic/identity.json` (or any path provided at
//! construction time) using the legacy JSON shape:
//!
//! ```json
//! {"secret":[...64 bytes...],"pubkey_base58":"..."}
//! ```
//!
//! `set` uses an atomic `NamedTempFile::new_in(parent) → persist()` write so
//! a crash mid-write never leaves a partially-written file.  Mode 0600 is
//! enforced on Unix.
use std::path::PathBuf;

use crate::identity::keystore::{KeyStore, KeystoreEntry, KeystoreError};

pub struct FileKeyStore {
    path: PathBuf,
}

impl FileKeyStore {
    /// Create a `FileKeyStore` operating on `path`.
    ///
    /// The directory containing `path` must exist before calling `set`.
    pub fn new(path: PathBuf) -> Self {
        Self { path }
    }
}

impl KeyStore for FileKeyStore {
    /// Read the entry from disk.
    ///
    /// Returns:
    /// - `Ok(Some(entry))` — legacy-shape file with `secret` key present.
    /// - `Ok(None)` — file absent, or file present but `secret` key missing
    ///   (stub shape). Callers should fall back to a different backend.
    /// - `Err(KeystoreError::Io(_))` — unexpected I/O error.
    /// - `Err(KeystoreError::Serde(_))` — file exists but is not valid JSON at
    ///   all (distinct from a stub-shape file).
    fn get(&self) -> Result<Option<KeystoreEntry>, KeystoreError> {
        let bytes = match std::fs::read(&self.path) {
            Ok(b) => b,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(None),
            Err(e) => return Err(KeystoreError::Io(e)),
        };

        // Parse as a generic JSON object first to detect stub shape (no "secret" key).
        let raw: serde_json::Value = serde_json::from_slice(&bytes)?;
        if raw.get("secret").is_none() {
            // Stub shape — secret lives in the OS keychain, not here.
            return Ok(None);
        }

        // Full legacy shape — deserialize into KeystoreEntry.
        let entry: KeystoreEntry = serde_json::from_value(raw)?;
        Ok(Some(entry))
    }

    /// Write the entry to disk atomically.
    ///
    /// Uses `NamedTempFile::new_in(parent)` + `.persist()` so a crash during
    /// the write never leaves the file in a partially-written state.  On Unix,
    /// the temp file inherits `NamedTempFile`'s default mode 0600, which is
    /// then preserved by the rename.
    fn set(&self, entry: &KeystoreEntry) -> Result<(), KeystoreError> {
        use tempfile::NamedTempFile;

        let parent = self.path.parent().ok_or_else(|| {
            KeystoreError::Io(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "identity path has no parent directory",
            ))
        })?;

        // Create a temp file in the same directory so the rename is atomic.
        let mut tmp = NamedTempFile::new_in(parent)?;

        // On Unix, enforce mode 0600 before writing.
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let perms = std::fs::Permissions::from_mode(0o600);
            std::fs::set_permissions(tmp.path(), perms)?;
        }

        // Write the JSON.  `serde_json::to_writer` is compact (no whitespace).
        serde_json::to_writer(&mut tmp, entry)?;

        // Atomically rename to the final path.
        tmp.persist(&self.path).map_err(|e| e.error)?;

        Ok(())
    }

    /// Delete the file if it exists.  No error if already absent.
    fn remove(&self) -> Result<(), KeystoreError> {
        match std::fs::remove_file(&self.path) {
            Ok(()) => Ok(()),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(e) => Err(KeystoreError::Io(e)),
        }
    }

    /// Returns `true` if the parent directory exists and is writable.
    fn available(&self) -> Result<bool, KeystoreError> {
        let parent = match self.path.parent() {
            Some(p) => p,
            None => return Ok(false),
        };
        // Probe write access by attempting a temp file creation.
        match tempfile::NamedTempFile::new_in(parent) {
            Ok(_) => Ok(true),
            Err(_) => Ok(false),
        }
    }

    fn name(&self) -> &'static str {
        "file"
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn make_entry(n: u8) -> KeystoreEntry {
        let mut secret = [0u8; 64];
        secret[0] = n;
        KeystoreEntry {
            secret,
            pubkey_base58: format!("pubkey_{n}"),
        }
    }

    fn store_in(dir: &TempDir) -> FileKeyStore {
        FileKeyStore::new(dir.path().join("identity.json"))
    }

    #[test]
    fn write_then_read_yields_same_entry() {
        let dir = TempDir::new().unwrap();
        let store = store_in(&dir);
        let entry = make_entry(99);
        store.set(&entry).unwrap();
        let got = store.get().unwrap().expect("entry should be readable");
        assert_eq!(got.secret, entry.secret);
        assert_eq!(got.pubkey_base58, entry.pubkey_base58);
    }

    #[cfg(unix)]
    #[test]
    fn mode_0600_on_unix() {
        use std::os::unix::fs::PermissionsExt;
        let dir = TempDir::new().unwrap();
        let store = store_in(&dir);
        store.set(&make_entry(1)).unwrap();
        let meta = std::fs::metadata(dir.path().join("identity.json")).unwrap();
        let mode = meta.permissions().mode() & 0o777;
        assert_eq!(mode, 0o600, "file mode must be 0600, got {mode:o}");
    }

    #[test]
    fn remove_deletes_file() {
        let dir = TempDir::new().unwrap();
        let store = store_in(&dir);
        store.set(&make_entry(2)).unwrap();
        store.remove().unwrap();
        assert!(!dir.path().join("identity.json").exists());
    }

    #[test]
    fn remove_on_missing_file_is_ok() {
        let dir = TempDir::new().unwrap();
        let store = store_in(&dir);
        // Remove without a prior set — must succeed.
        store.remove().unwrap();
    }

    #[test]
    fn get_on_missing_returns_ok_none() {
        let dir = TempDir::new().unwrap();
        let store = store_in(&dir);
        let got = store.get().unwrap();
        assert!(got.is_none());
    }

    #[test]
    fn get_on_stub_shape_returns_ok_none() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("identity.json");
        // Write a stub (no "secret" key) directly.
        let stub = r#"{"pubkey_base58":"abc123","keychain_ref":"xyz.mnemonik.identity/default","created_at":"2026-01-01T00:00:00Z"}"#;
        std::fs::write(&path, stub).unwrap();

        let store = FileKeyStore::new(path);
        let got = store.get().unwrap();
        assert!(
            got.is_none(),
            "stub-shaped file must return Ok(None), not an entry"
        );
    }

    #[test]
    fn available_returns_true_for_writable_dir() {
        let dir = TempDir::new().unwrap();
        let store = store_in(&dir);
        assert!(store.available().unwrap());
    }
}

File: packages/cli/src/identity/keystore-file.ts
/**
 * FileKeyStore — filesystem-backed KeyStore implementation.
 *
 * Reads and writes a JSON file at a caller-supplied path using the canonical
 * shape: {"secret":[...64 bytes...],"pubkey_base58":"..."}.
 *
 * `set` uses an atomic tempfile-then-rename pattern so a crash mid-write never
 * leaves a partially-written file at the target path.  Mode 0o600 is enforced
 * on Unix via writeFile options.  On Windows, fs.chmod is a no-op — this is a
 * documented limitation; the ACL helper from mnemonic-cli (if present) could
 * be called here in a future iteration.
 *
 * SECURITY: `secret` bytes are NEVER logged, included in error messages, or
 * placed in Error.cause in any retrievable form.
 */

import fs from "node:fs/promises";
import path from "node:path";
import { randomBytes } from "node:crypto";
import {
  type KeyStore,
  type KeystoreEntry,
  KeystoreError,
  canonicalEntryJson,
} from "./keystore.js";

export class FileKeyStore implements KeyStore {
  public readonly name = "file" as const;

  constructor(private readonly filePath: string) {}

  /**
   * Read the entry from disk.
   *
   * Returns:
   * - entry  — full shape with `secret` present.
   * - null   — file absent, or stub shape (no `secret` key).
   * Throws:
   * - KeystoreError('io')    — unexpected I/O error.
   * - KeystoreError('serde') — file is not valid JSON.
   */
  async get(): Promise<KeystoreEntry | null> {
    let raw: string;
    try {
      raw = await fs.readFile(this.filePath, "utf8");
    } catch (err) {
      if (isEnoent(err)) return null;
      throw new KeystoreError(
        "io",
        `FileKeyStore.get: cannot read file (path length=${this.filePath.length})`,
        // Omit cause — it contains the fs path which is benign but not the
        // secret, so technically fine; we omit it for uniformity.
      );
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(raw);
    } catch {
      throw new KeystoreError(
        "serde",
        "FileKeyStore.get: file is not valid JSON",
      );
    }

    if (typeof parsed !== "object" || parsed === null) {
      throw new KeystoreError(
        "serde",
        "FileKeyStore.get: top-level value is not an object",
      );
    }

    const p = parsed as Record<string, unknown>;

    // Stub shape — secret lives in OS keychain, not in this file.
    if (!("secret" in p)) {
      return null;
    }

    if (!Array.isArray(p["secret"])) {
      throw new KeystoreError(
        "serde",
        "FileKeyStore.get: 'secret' is not an array",
      );
    }
    if (typeof p["pubkey_base58"] !== "string") {
      throw new KeystoreError(
        "serde",
        "FileKeyStore.get: 'pubkey_base58' is not a string",
      );
    }

    return {
      secret: Array.from(p["secret"] as ArrayLike<number>),
      pubkey_base58: p["pubkey_base58"] as string,
    };
  }

  /**
   * Write the entry atomically.
   *
   * Writes to `<path>.tmp.<random>` with mode 0o600, then renames to the
   * final path.  The rename is atomic on POSIX (same filesystem).  On Windows,
   * rename may fail if the target exists in some Node versions — we handle
   * that by catching EPERM and retrying with unlink+rename.
   *
   * NOTE: fs.chmod is a no-op on Windows.  The 0o600 mode in writeFile options
   * is equally a no-op there.  This is a known limitation; Windows ACLs require
   * the win-acl helper (future iteration).
   */
  async set(entry: KeystoreEntry): Promise<void> {
    const json = canonicalEntryJson(entry);
    const suffix = randomBytes(4).toString("hex");
    const tmpPath = this.filePath + ".tmp." + suffix;

    try {
      // Write to temp file with mode 0o600 (no-op on Windows — see above).
      await fs.writeFile(tmpPath, json, { encoding: "utf8", mode: 0o600 });
    } catch (err) {
      throw new KeystoreError(
        "io",
        "FileKeyStore.set: cannot write temporary file",
      );
    }

    try {
      await fs.rename(tmpPath, this.filePath);
    } catch (err) {
      // On Windows, rename over an existing file may throw EPERM.
      // Retry with unlink + rename.
      if (isEperm(err)) {
        try {
          await fs.unlink(this.filePath);
          await fs.rename(tmpPath, this.filePath);
        } catch {
          // Best-effort cleanup of the temp file; ignore secondary errors.
          void fs.unlink(tmpPath).catch(() => undefined);
          throw new KeystoreError(
            "io",
            "FileKeyStore.set: atomic rename failed (Windows EPERM fallback also failed)",
          );
        }
      } else {
        // Clean up the temp file before surfacing the error.
        void fs.unlink(tmpPath).catch(() => undefined);
        throw new KeystoreError("io", "FileKeyStore.set: atomic rename failed");
      }
    }
  }

  /** Delete the file. Idempotent — no error if already absent. */
  async remove(): Promise<void> {
    try {
      await fs.unlink(this.filePath);
    } catch (err) {
      if (isEnoent(err)) return;
      throw new KeystoreError("io", "FileKeyStore.remove: unlink failed");
    }
  }

  /**
   * Return true if the parent directory is writable.
   * Does NOT create the directory — callers (e.g. ensure()) must do that.
   */
  async available(): Promise<boolean> {
    const parent = path.dirname(this.filePath);
    try {
      await fs.access(parent, fs.constants.W_OK);
      return true;
    } catch {
      return false;
    }
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isEnoent(err: unknown): boolean {
  return (
    typeof err === "object" &&
    err !== null &&
    "code" in err &&
    (err as NodeJS.ErrnoException).code === "ENOENT"
  );
}

function isEperm(err: unknown): boolean {
  return (
    typeof err === "object" &&
    err !== null &&
    "code" in err &&
    (err as NodeJS.ErrnoException).code === "EPERM"
  );
}
